### DISCUSSION TITLE
Unable to identify upstream interface

### DISCUSSION THREAD

**Discussion by zll123dxd** [Mygod/VPNHotspot#319] [Q&A] [answered]
Are Settings> VPN, tun on ppp0 and 3rd party vpn different?
I found that they apply different iptables rules, but only use 3rd party app to surf the internet.

Please tell me 
Thanks

**Comment by Mygod** [comment, marked as answer]
It should work for system VPN as well.

**Comment by zll123dxd** [comment]
I'm sorry, but  it not work normal.
It's ok for me to use another phone.

ipconfig :
rmnet_data2  - 10.24.240.109
wlan0 -192.168.43.1
ppp0 -  inet addr:192.168.61.9  P-t-P:192.168.60.243

Referring to another phone, I used the following rules:
adb shell ip rule add  iif wlan0 unreachable priority 17980
adb shell ip rule add  iif wlan0 lookup rmnet_data2 priority 17900
adb shell ip rule add  iif wlan0 lookup ppp0 priority 17800
adb shell ip rule add to 192.168.60.243 iif wlan0 lookup ppp0 priority 17700

adb shell iptables -w -N vpnhotspot_fwd
adb shell iptables -w -N vpnhotspot_acl
adb shell iptables -w -t filter -I FORWARD -j vpnhotspot_fwd
adb shell iptables -w -t filter -I vpnhotspot_fwd -i wlan0 -j vpnhotspot_acl
adb shell iptables -w -t filter -I vpnhotspot_fwd -o wlan0 -m state --state ESTABLISHED,RELATED -j vpnhotspot_acl
adb shell iptables -w -t filter -A vpnhotspot_fwd -i wlan0 ! -o wlan0 -j REJECT
adb shell iptables -w -t nat -N vpnhotspot_masquerade
adb shell iptables -w -t nat -I POSTROUTING -j vpnhotspot_masquerade
adb shell iptables -w -t nat -A vpnhotspot_masquerade -s 192.168.43.1/24 -o rmnet_data2 -j MASQUERADE
adb shell iptables -w -t nat -A vpnhotspot_masquerade -s 192.168.43.1/24 -o ppp0 -j MASQUERADE
adb shell iptables -w -t nat -A PREROUTING -i wlan0 -p tcp -d 192.168.43.1 --dport 53 -j DNAT --to-destination 192.168.60.243
adb shell iptables -w -t nat -A PREROUTING -i wlan0 -p udp -d 192.168.43.1 --dport 53 -j DNAT --to-destination 192.168.60.243 (p2p server)
adb shell iptables -w -t filter -I vpnhotspot_acl -i wlan0 -s 192.168.43.40 -j ACCEPT
adb shell iptables -w -t filter -I vpnhotspot_acl -o wlan0 -d 192.168.43.40 -j ACCEPT  (192.168.43.40 is client Device IP)

I use two different mobile phones to apply the above rules, both are ok.

But the device I want the use to not work on.
What are the possible reasons for me to check?

**Comment by zll123dxd** [comment]
adb shell ip rule add  iif wlan0 unreachable priority 17980
adb shell ip rule add  iif wlan0 lookup rmnet_data2 priority 17900
adb shell ip rule add  iif wlan0 lookup ppp0 priority 17800
adb shell ip rule add to 192.168.60.243 iif wlan0 lookup ppp0 priority 17700
#p2p server ip

adb shell iptables -w -N vpnhotspot_fwd
adb shell iptables -w -t filter -I FORWARD -j vpnhotspot_fwd
adb shell iptables -w -t filter -I vpnhotspot_fwd -i wlan0 -j ACCEPT
adb shell iptables -w -t filter -I vpnhotspot_fwd -o wlan0 -m state --state ESTABLISHED,RELATED -j ACCEPT
adb shell iptables -w -t filter -A vpnhotspot_fwd -i wlan0 ! -o wlan0 -j REJECT

adb shell iptables -w -t nat -N vpnhotspot_masquerade
adb shell iptables -w -t nat -I POSTROUTING -j vpnhotspot_masquerade
adb shell iptables -w -t nat -A vpnhotspot_masquerade -s 192.168.43.1/24 -o ppp0 -j MASQUERADE
adb shell iptables -w -t nat -A PREROUTING -i wlan0 -p tcp -d 192.168.43.1 --dport 53 -j DNAT --to-destination 192.168.60.243
adb shell iptables -w -t nat -A PREROUTING -i wlan0 -p udp -d 192.168.43.1 --dport 53 -j DNAT --to-destination 192.168.60.243

I simplified the command again. It still works fine.
But still not working on my android 10 device.

**Comment by Mygod** [comment]
If it only does not work on a specific phone, then you can try to report to that phone/ROM developer, but you will most likely have to do some troubleshooting. A good place to start is maybe to do some tcpdump perhaps.
